docs/operations/problem-determination.md

```diff
@@ -0,0 +1,223 @@
+  * enable the [heap profiler](https://nodejs.org/dist/latest-v17.x/docs/api/cli.html#--heap-prof)
+    and look at where allocations are taking place
+
+Generating heap dumps will have a performance impact on the process both in terms of memory
+and execution when the dump is being generated. Some of the team's suggestions for limiting the impact:
+* only enable heap dumps for one process/instance of the application, leaving the others to better server customers
+* try to minimize the size of the heap dumps. For example, if you increased the memory allocated to the application
+  as part of trying to investigate the issue, revert that change. If possible limit the instance of the application
+  to using less memory than you normally would.
+* make sure you have enough additional memory available on the machine running the process. The dump may need double
+  the size of heap while it is being generated.
+* make sure to disable Kubernetes liveness checks (may require new deployment with that setting) or
+  the periodic liveness check may kill your app before it finishes creating the heap dump. 
+
+#### Hangs or slow performance
+
+* Symptoms
+  * process health checks fail
+  * metrics report longer than expected request times
+  * metrics report higher than expected CPU usage
+
+* Approach
+  * generate a [Diagnostic report](https://nodejs.org/api/report.html) and look for ref flags
+  * generate a [flame graph](https://nodejs.org/en/docs/guides/diagnostics-flamegraph/). The team
+    has had success with [0x](https://www.npmjs.com/package/0x) and [bubbleprof(https://clinicjs.org/bubbleprof/)
+    and [Flame](https://clinicjs.org/flame/)
+
+It may be difficult to investigate performance issues in the production system. Once you have identified
+that there is a problem, being able to recreate in another environment make it much easier. The team has
+had success using [goreplay](https://github.com/buger/goreplay) as well as
+[Autocannon](https://www.npmjs.com/package/autocannon) to generate load in test environments in order to
+reproduce issue seen in production.
+
+If you must investigate in production, you may have to disable or extend the health check period so that
+you can generate diagnostic reports and/or flame graphs. Ensure that you don't allow more than one
+process to be in the hung/slow state while investigating in order to limit impact on end users.
+
+#### Application failures
+
+* Symtoms
+  * metrics report failure reponses
+
+* Approach
+  * Review application logs, turn on additional levels of logging if necessary
+  * Review [transaction traces](https://github.com/nodeshift/nodejs-reference-architecture/blob/main/docs/operations/distributed-tracing.md)
+
+
+#### Unhandled promise rejections or exceptions
+
+* Symtoms
+  * process terminatese reporting an unhandled rejection or exception
+  
+* Approach
+  * review log to find stack strace, review code indicated in stack trace
```

Comment:
im not sure if we have mentioned it anywhere in our reference architecture, but stacktraces and sourcemaps might be something to mention, otherwise stacktraces might not be useful since it could be obfuscated.